Gutenblocks: Add VR block

ockham opened the pull request:
[Moved](https://github.com/Automattic/jetpack/blob/ff229729399161b6ed0ff8b19b05e30a55928b2d/modules/shortcodes/js/blocks/vr-block.jsx) from the Jetpack repo. Some modifications to be consistent with Gutenblocks in Calypso.

I'm not yet adding this to the `jetpack` preset, since it's not in good enough shape yet, so testing requires a bit more manual work.

### Testing Instructions

In your local Calypso repo:
- Checkout this branch
- `npm run sdk -- gutenberg client/gutenberg/extensions/vr/`

In your local JP repo:
- Checkout the `remove/vr-gutenblock` branch (Automattic/jetpack#10331)
- Run `yarn build`
- Delete `_inc/blocks/editor.*` and `_inc/blocks/view.*`
- Copy the contents of your Calypso repo's `client/gutenberg/extensions/vr/build` to your `_inc/blocks`
- `yarn docker:up`
- (Log into `localhost/wp-admin`, connect Jetpack to WP.com)
- Add a VR block to a post or page (you can find it under "embeds")
- The block should work in different screen sizes
- Adding a valid URL _and_ type will render the embed. Here's a test URL you can use together with "360" setting: `https://en-blog.files.wordpress.com/2016/12/regents_park.jpg`

simison commented on client/gutenberg/extensions/vr/editor.js:
```diff
@@ -0,0 +1,107 @@
+ */
+import './editor.scss';
+
+registerBlockType( 'jetpack/vr', {
```
Let's rename this to `a8c/vr` like rest of the blocks?

tyxla commented on client/gutenberg/extensions/vr/editor.js:
```diff
@@ -0,0 +1,107 @@
+		},
+	},
+
+	edit: props => {
```
Maybe the `edit` would make sense in a separate component+file.

tyxla commented on client/gutenberg/extensions/vr/editor.js:
```diff
@@ -0,0 +1,107 @@
+				>
+					<TextControl
+						type="url"
+						style={ { flex: '1 1 auto' } }
```
Could we define those in a stylesheet file instead of having to specify styles inline?

ockham replied:
Removing it altogether since it [doesn't seem to have any effect](https://github.com/Automattic/jetpack/pull/9935#discussion_r204074342).

tyxla commented on client/gutenberg/extensions/vr/editor.js:
```diff
@@ -0,0 +1,107 @@
+			</div>
+		);
+	},
+	save: props => {
```
I think `save` would make sense in a separate component and file too.

tyxla commented on client/gutenberg/extensions/vr/editor.js:
```diff
@@ -0,0 +1,107 @@
+			);
+		}
+		return (
+			<div>
```
Use a `<Fragment />` instead?

tyxla commented on client/gutenberg/extensions/vr/editor.js:
```diff
@@ -0,0 +1,107 @@
+	attributes: {
+		url: {
+			type: 'string',
+			default: '',
```
Since we specify empty strings for defaults, are those defaults even necessary?

tyxla commented on client/gutenberg/extensions/vr/editor.js:
```diff
@@ -0,0 +1,107 @@
+
+registerBlockType( 'jetpack/vr', {
+	title: __( 'VR Image', 'jetpack' ),
+	description: __( 'Embed 360° photos and Virtual Reality (VR) Content', 'jetpack' ),
```
Uppercase in "Virtual Reality (VR) Content" necessary?

tyxla commented on client/gutenberg/extensions/vr/editor.js:
```diff
@@ -0,0 +1,107 @@
+	title: __( 'VR Image', 'jetpack' ),
+	description: __( 'Embed 360° photos and Virtual Reality (VR) Content', 'jetpack' ),
+	icon: 'embed-photo',
+	category: 'embed',
```
Do we want the `Jetpack` category for this one too?

tyxla commented on client/gutenberg/extensions/vr/editor.js:
```diff
@@ -0,0 +1,107 @@
+			type: 'string',
+			default: '',
+		},
+	},
```
This one could also use some keywords. Maybe the `embed` from above could fit well here.

tyxla replied:
Other suggestions could be `360`, `vr` or `panorama`

tyxla commented on client/gutenberg/extensions/vr/editor.js:
```diff
@@ -0,0 +1,107 @@
+	edit: props => {
+		const { attributes, setAttributes } = props;
+
+		const onChangeUrl = value => setAttributes( { url: value.trim() } );
```
The fact that we define these in every call is a great reason to have this in a separate component.

tyxla commented on client/gutenberg/extensions/vr/editor.js:
```diff
@@ -0,0 +1,107 @@
+	save: props => {
+		return (
+			<div className={ props.className }>
+				<iframe
```
Could we split out this `iframe` in a component? It seems it can be made reusable.

tyxla commented on client/gutenberg/extensions/vr/editor.scss:
```diff
@@ -0,0 +1,20 @@
+.wp-block-jetpack-vr {
+	position: relative;
+	max-width: 525px;
```
I wonder where this value comes from.

ockham replied:
I think it's pretty arbitrary. Inclined to leave it for now.

tyxla commented on client/gutenberg/extensions/vr/editor.js:
```diff
@@ -0,0 +1,107 @@
+/** @format */
```
Should we add this block to the `jetpack` preset so we can load it?

ockham replied:
I didn't add it deliberately:

> I'm not yet adding this to the `jetpack` preset, since it's not in good enough shape yet [...]

While I wanted to remove it from the JP repo, I didn't want to commit us to polishing it for inclusion with Jetpack, since other blocks have higher priority.

ockham commented:
Thanks @tyxla! I didn't fix anything rather than the most basic Calypso-style `import` since I didn't want to spend too much time on it. Your suggestions make sense tho, and most of them should be easy wins, so I'll address them.

ockham commented:
I think I've addressed all feedback. Note that since the block itself doesn't declare the `jetpack` category, you might need to revert the category to `embed` to test the block:

```diff
diff --git a/client/gutenberg/extensions/vr/editor.js b/client/gutenberg/extensions/vr/editor.js
index 910f608105..ebba95138c 100644
--- a/client/gutenberg/extensions/vr/editor.js
+++ b/client/gutenberg/extensions/vr/editor.js
@@ -16,7 +16,7 @@ registerBlockType( 'a8c/vr', {
        title: __( 'VR Image', 'jetpack' ),
        description: __( 'Embed 360° photos and Virtual Reality (VR) content', 'jetpack' ),
        icon: 'embed-photo',
-       category: 'jetpack',
+       category: 'embed',
        keywords: [ __( 'vr', 'jetpack' ), __( 'panorama', 'jetpack' ), __( '360', 'jetpack' ) ],
        support: {
                html: false,
```

sirreal commented:
Why don't we include this in the preset?

ockham commented:
> Why don't we include this in the preset?

I don't think it's in good enough shape yet, and I didn't want to commit us to giving it a lot of attention (yet) when there are higher priority blocks.

I'm starting to consider adding it now tho, since @tyxla was asking the same thing :thinking:

tyxla commented:
Yeah, it's a good reason to keep it away from the preset for now 👍

sirreal commented:
Was it available in Jetpack before by default?

I agree that it may not be ready for primetime, especially as Gutenberg adoption is expected to truly soar in the near future 🚀 😉 

I do want to understand whether we're removing something that was previously available.

tyxla commented on client/gutenberg/extensions/vr/save.jsx:
```diff
@@ -0,0 +1,23 @@
+			width="100%"
+			height="300"
+			src={
+				'https://vr.me.sh/view/?view=' +
```
Gutenberg has a nice `addQueryArgs` utility we can use for building the query string - https://github.com/WordPress/gutenberg/blob/adb1f0e6e8271ce2b9c49963444add331ae90017/packages/url/src/index.js#L29

ockham replied:
#27929

simison commented:
>Was it available in Jetpack before by default?

Yes (still is in latest release), but it doesn't work on Gutenberg 3.2.0+

ockham commented:
Thanks all!

ockham commented:
> Finally: let's make sure to add a `Co-authored-by:` in the commit to make sure we give credit to the users who worked on the original block in Jetpack.

Argh now I did forget to do that :man_facepalming:
